README.md
qps
=======

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![Gittip][gittip-image]][gittip-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/qps.svg?style=flat-square
[npm-url]: https://npmjs.org/package/qps
[travis-image]: https://img.shields.io/travis/node-modules/qps.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/qps
[coveralls-image]: https://img.shields.io/coveralls/node-modules/qps.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/node-modules/qps?branch=master
[gittip-image]: https://img.shields.io/gittip/fengmk2.svg?style=flat-square
[gittip-url]: https://www.gittip.com/fengmk2/
[david-image]: https://img.shields.io/david/node-modules/qps.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/qps
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/qps.svg?style=flat-square
[download-url]: https://npmjs.org/package/qps

High performance qps counter

## Install

```bash
$ npm install qps
```

## Usage

```js
var QPSCounter = require('qps');

var counter = new QPSCounter();

// plus one
counter.plus();

// current qps
counter.get();
```

## API

### QPSCounter(options)

Create a new qps counter with `options`.

- options.listener: a `Function(qpsList)` to get one minute before qps list

```js
var counter = new QPSCounter({
  // get qps list every minute
  listener: function (qpsList) {
    console.log('current process handler qps: %j', qpsList);
    // [85,82,86,90,87,87,88,87,...,88]
  }
});

// plus one when new request arrive
counter.plus();
```

### #plus([count])

Plus one request

```js
// plus one
counter.plus();

// plus more
counter.plus(100);
```

### #get()

Get current requests per second (QPS)

```js
console.log(counter.get());
```

## License

(The MIT License)

Copyright (c) 2014 fengmk2 <fengmk2@gmail.com> and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

1.1.0 / 2014-11-14
==================

 * refactor: get() meaning plus(0)

1.1.0 / 2014-11-13
==================

 * plus support number as input

1.0.1 / 2014-11-12
==================

 * plus and return the current qps

1.0.0 / 2014-11-11
==================

 * feat: can get every minute qps counts


test/log.js
/**!
 * qps - test/log.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var qps = require('../');

var counter = qps({
  listener: total
});

function request() {
  counter.plus();
}
setInterval(request, 10);

function total(qpsList) {
  var now = new Date();
  var index = now.getMinutes() % 2;
  console.log('%s, %s:\nprev   : %j\ncurrent: %j',
    now, Date.now(), qpsList, counter.counts[index]);
}

// $ node test/log.js
// Tue Nov 11 2014 14:07:21 GMT+0800 (CST), 1415686041867:
// prev   : [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,13,85,85,86,86,86,85,89,88,86,87,85,86,84,85,88,87,88,84,87,86,87,89,89,86,88,89,87,86,89,89,88,91,88,88,87,88,84,85]
// current: [85,82,86,90,87,87,88,87,85,87,86,84,87,87,85,82,84,85,85,86,85,72,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:08:21 GMT+0800 (CST), 1415686101895:
// prev   : [85,82,86,90,87,87,88,87,85,87,86,84,87,87,85,82,84,85,85,86,85,84,88,86,85,85,84,84,85,84,84,85,82,84,80,81,83,82,86,81,82,84,83,88,83,84,83,83,84,84,85,87,87,86,88,85,85,86,86,90]
// current: [87,88,86,88,86,84,87,87,85,85,86,85,86,86,85,88,85,85,88,88,87,77,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:09:21 GMT+0800 (CST), 1415686161902:
// prev   : [87,88,86,88,86,84,87,87,85,85,86,85,86,86,85,88,85,85,88,88,87,86,87,86,86,87,88,86,86,83,81,83,88,85,87,86,87,86,87,88,85,87,86,87,85,84,87,81,88,87,86,88,89,89,87,84,85,86,87,86]
// current: [87,89,88,86,87,87,90,87,87,87,87,89,87,87,85,83,84,85,87,86,86,79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:10:21 GMT+0800 (CST), 1415686221907:
// prev   : [87,89,88,86,87,87,90,87,87,87,87,89,87,87,85,83,84,85,87,86,86,87,87,87,89,84,87,87,87,86,87,87,87,85,84,85,87,84,85,86,85,86,86,87,89,88,88,87,87,87,88,88,88,88,88,87,85,87,87,86]
// current: [89,89,86,87,85,86,87,85,87,83,86,88,86,84,86,88,86,85,87,84,87,78,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:11:21 GMT+0800 (CST), 1415686281909:
// prev   : [89,89,86,87,85,86,87,85,87,83,86,88,86,84,86,88,86,85,87,84,87,86,88,85,86,88,84,90,85,88,86,84,88,86,84,86,86,85,89,86,89,85,89,87,86,84,84,84,86,85,86,87,87,85,86,86,85,83,85,85]
// current: [87,89,86,85,86,87,85,84,84,85,84,84,85,85,83,85,86,85,87,86,87,78,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:12:21 GMT+0800 (CST), 1415686341914:
// prev   : [87,89,86,85,86,87,85,84,84,85,84,84,85,85,83,85,86,85,87,86,87,86,88,87,85,85,84,87,84,85,86,85,85,86,83,85,85,86,84,85,85,83,86,85,89,85,84,86,84,87,85,84,86,85,85,85,84,86,85,85]
// current: [86,88,88,87,85,86,84,87,85,85,84,86,86,87,85,86,85,86,84,87,87,78,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
// Tue Nov 11 2014 14:13:21 GMT+0800 (CST), 1415686401917:
// prev   : [86,88,88,87,85,86,84,87,85,85,84,86,86,87,85,86,85,86,84,87,87,86,88,84,86,85,84,84,88,85,84,85,85,85,86,85,86,85,85,84,85,84,86,85,85,86,84,84,85,85,88,85,85,86,86,85,85,85,84,86]


test/qps.test.js
/**!
 * qps - test/qps.test.js
 *
 * Copyright(c) 2014 fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

"use strict";

/**
 * Module dependencies.
 */

var qps = require('../');

describe('qps.test.js', function () {
  it('should plus work', function (done) {
    var counter = qps();
    counter.plus().should.equal(1);
    counter.get().should.equal(1);
    counter.plus().should.equal(2);

    counter.get().should.equal(2);
    counter.get().should.equal(2);
    counter.get().should.equal(2);
    counter.get().should.equal(2);
    setTimeout(function () {
      counter.get().should.equal(0);
      counter.plus();
      counter.get().should.equal(1);
      counter.get().should.equal(1);
      counter.get().should.equal(1);
      counter.get().should.equal(1);
      counter.plus();
      counter.plus();
      counter.plus(2).should.equal(5);
      counter.get().should.equal(5);
      done();
    }, 1000);
  });

  it('should get one minute before qps', function () {
    var counter = qps();
    counter.plus();
    counter.plus();
    var qpsList = counter.listAndResetOneMinuteBefore();
    qpsList.should.length(60);
  });

  it('should init with options.listener', function (done) {
    var counter = qps({
      listener: function (qpsList) {
        qpsList.should.length(60);
        counter.close();
        // close again should work
        counter.close();
        done();
      }
    });
    counter.plus();
    counter.plus();
    setTimeout(function () {
      counter._onOneMinute();
    }, 100);
  });
});


index.js
/**!
 * qps - index.js
 *
 * Copyright(c) 2014 fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

"use strict";

/**
 * Module dependencies.
 */

module.exports = QPSCounter;

function QPSCounter(options) {
  if (!(this instanceof QPSCounter)) {
    return new QPSCounter(options);
  }

  this.ts = [[], []];
  this.counts = [[], []];
  this.tmpCounts = [];
  for (var i = 0; i < 60; i++) {
    this.counts[0][i] = 0;
    this.ts[0][i] = 0;
    this.counts[1][i] = 0;
    this.ts[1][i] = 0;
    this.tmpCounts[i] = 0;
  }

  this.timer;
  this.listener = options && options.listener;
  // listener for one minute before qps
  // listener format: `Function listener(qpsList)`
  if (this.listener) {
    this.timer = setInterval(this._onOneMinute.bind(this), 60000);
  }
}

var proto = QPSCounter.prototype;

proto.plus = function (count) {
  return this.get(count || 1);
};

proto.get = function (count) {
  var now = new Date();
  var index = now.getMinutes() % 2;
  var second = now.getSeconds();
  var now = Date.now();
  if (now - this.ts[index][second] > 2000) {
    this.ts[index][second] = now;
    this.counts[index][second] = 0;
  }
  if (count) {
    return (this.counts[index][second] += count);
  }
  return this.counts[index][second];
};

proto.close = function () {
  if (this.timer) {
    clearInterval(this.timer);
    this.timer = null;
  }
};

proto.listAndResetOneMinuteBefore = function () {
  var now = new Date();
  var index = now.getMinutes() % 2;
  if (index === 0) {
    index = 1;
  } else {
    index = 0;
  }
  var cs = this.counts[index];
  for (var i = 0; i < cs.length; i++) {
    this.tmpCounts[i] = cs[i];
    cs[i] = 0;
  }
  return this.tmpCounts;
};

proto._onOneMinute = function () {
  this.listener(this.listAndResetOneMinuteBefore());
};


benchmark/qps.js
/**!
 * qps - benchmark/qps.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var Benchmark = require('benchmark');
var benchmarks = require('beautify-benchmark');
var qps = require('../');

var suite = new Benchmark.Suite();

var counter = qps();

// add tests
suite

.add("counter.plus()", function() {
  counter.plus();
})
.add("counter.get()", function() {
  counter.get();
})
.add("counter.plus(), counter.get()", function() {
  counter.plus();
  counter.get();
})
.add("counter.listAndResetOneMinuteBefore()", function() {
  counter.listAndResetOneMinuteBefore();
})

// add listeners
.on('cycle', function (event) {
  benchmarks.add(event.target);
})
.on('start', function () {
  console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
})
.on('complete', function() {
  benchmarks.log();
})
// run async
.run({ 'async': false });

// $ node benchmark/qps.js
//
//
//   node version: v0.11.14, date: Tue Nov 11 2014 15:54:16 GMT+0800 (CST)
//   Starting...
//   4 tests completed.
//
//   counter.plus()                        x 3,117,962 ops/sec ±1.97% (91 runs sampled)
//   counter.get()                         x 3,105,478 ops/sec ±2.65% (90 runs sampled)
//   counter.plus(), counter.get()         x 1,637,585 ops/sec ±1.13% (90 runs sampled)
//   counter.listAndResetOneMinuteBefore() x 3,179,284 ops/sec ±1.43% (90 runs sampled)

// $ node benchmark/qps.js
//
//
//   node version: v0.11.14, date: Tue Nov 11 2014 03:18:36 GMT+0800 (CST)
//   Starting...
//   3 tests completed.
//
//   counter.plus()                x 3,382,367 ops/sec ±1.23% (94 runs sampled)
//   counter.get()                 x 3,291,861 ops/sec ±2.51% (90 runs sampled)
//   counter.plus(), counter.get() x 1,718,300 ops/sec ±1.10% (93 runs sampled)
